### PR TITLE
Support grant statement

### PR DESCRIPTION
**Pull request by shenli**
https://dev.mysql.com/doc/refman/5.7/en/grant.html
Support basic grant syntax, including global/db/table/column scope privilege granting. Each scope has a few unittest case.

### CODE REVIEW

**Comment by ngaut** [comment]
WOW,this is a huge PR.

File: bootstrap.go [hunk 1]
```diff
@@ -43,6 +43,7 @@ const (
 		Alter_priv   ENUM('N','Y') NOT NULL  DEFAULT 'N',
 		Show_db_priv ENUM('N','Y') NOT NULL  DEFAULT 'N',
 		Execute_priv ENUM('N','Y') NOT NULL  DEFAULT 'N',
+		Index_priv ENUM('N','Y') NOT NULL  DEFAULT 'N',
```
Comment (qiuyesuifeng) [inline bootstrap.go]: Need alignment?

File: parser/parser.y [hunk 2]
```diff
@@ -4030,10 +4041,14 @@ CreateUserStmt:
 UserSpecification:
 	Username AuthOption	
 	{
-		$$ = &coldef.UserSpecification{
+		x := &coldef.UserSpecification{
```
Comment (qiuyesuifeng) [inline parser/parser.y]: Seems old code is OK?
If $2 == nil, also x.AuthOpt will be nil?

Comment (shenli) [inline parser/parser.y]: Yes, $2 maybe nil and you will get error.

File: bootstrap.go (hunk 1 above)
Comment (shenli) [inline bootstrap.go]: This is sql statement text, not golang code. But be alignment is better, I will fix this.

File: parser/parser.y [hunk 3]
```diff
@@ -4064,5 +4081,149 @@ AuthOption:
...
+		$$ = &coldef.GrantLevel {
+			Level: coldef.GrantLevelDB,
+		}		
+	}
+|	'*' '.' '*'
+	{
+		$$ = &coldef.GrantLevel {
+			Level: coldef.GrantLevelGlobal,
+		}		
+
```
Comment (qiuyesuifeng) [inline parser/parser.y]: Remove empty line.

File: stmt/stmts/grant.go [hunk 4]
```diff
@@ -0,0 +1,760 @@
...
+		cPriv = columnPriv.Name
+	}
+	return cPriv, nil
+}
+
+// Find the schema by dbName.
+func (s *GrantStmt) getTargetSchema(ctx context.Context) (*model.DBInfo, error) {
+	dbName := s.Level.DBName
+	if len(dbName) == 0 {
+		// Grant *, user current schema
```
Comment (qiuyesuifeng) [inline stmt/stmts/grant.go]: use current schema?

File: stmt/stmts/grant.go [hunk 5]
```diff
@@ -0,0 +1,760 @@
...
+		dbName = db.GetCurrentSchema(ctx)
+	}
+	if len(dbName) == 0 {
+		return nil, errors.Errorf("Miss DB name in grant db scope privilege.")
```
Comment (qiuyesuifeng) [inline stmt/stmts/grant.go]: Use errors.New,  and can make sure it belong to grant db scope?

File: stmt/stmts/grant.go [hunk 6]
```diff
@@ -0,0 +1,760 @@
...
+		// Grant each priv to the user.
+		for _, priv := range s.Privs {
+			if len(priv.Cols) > 0 {
+				// Check column scope privilege entry.
+				err1 := s.checkAndInitColumnPriv(ctx, userName, host, priv.Cols)
+				if err1 != nil {
+					return nil, errors.Trace(err1)
+				}
+			}
+			err2 := s.grantPriv(ctx, priv, user)
```
Comment (qiuyesuifeng) [inline stmt/stmts/grant.go]: Any need to check grant validity first? like grant xxx, then we will not to insert rows above.

File: stmt/stmts/grant.go [hunk 7]
```diff
@@ -0,0 +1,760 @@
...
+		if !ok {
+			return nil, errors.Errorf("Unknown priv: %s", priv)
+		}
+		if len(currTablePriv) == 0 {
+			newTablePriv = p
+		} else {
+			newTablePriv = fmt.Sprintf("%s,%s", currTablePriv, p)
+		}
+		for _, cp := range mysql.AllColumnPrivs {
+			if priv == cp {
```
Comment (qiuyesuifeng) [inline stmt/stmts/grant.go]: I see u use `mysql.Priv2SetStr[priv]` above, but why here use equal for check?

File: stmt/stmts/grant.go (hunk 6 above)
Comment (shenli) [inline stmt/stmts/grant.go]: Maybe later.

File: stmt/stmts/grant.go (hunk 7 above)
Comment (shenli) [inline stmt/stmts/grant.go]: p is string but cp and priv is PrivilegeType.
p is used to modify TypeSet column.

**Comment by shenli** [comment]
PTAL

File: stmt/stmts/grant.go [hunk 8]
```diff
@@ -0,0 +1,761 @@
...
+func initDBPrivEntry(ctx context.Context, user string, host string, db string) error {
+	st := &InsertIntoStmt{
+		TableIdent: table.Ident{
+			Name:   model.NewCIStr(mysql.DBTable),
+			Schema: model.NewCIStr(mysql.SystemDB),
+		},
+		ColNames: []string{"Host", "User", "DB"},
+	}
+	values := make([][]expression.Expression, 0, 1)
+	value := make([]expression.Expression, 0, 3)
```
Comment (c4pt0r) [inline stmt/stmts/grant.go]: value := []expression.Expression{
   &expression.Value{Val: host},
   &expression.Value{Val: user},
   &expression.Value{Val: db},
}

it's neater.

File: stmt/stmts/grant.go [hunk 9]
```diff
@@ -0,0 +1,761 @@
...
+func initTablePrivEntry(ctx context.Context, user string, host string, db string, tbl string) error {
+	st := &InsertIntoStmt{
+		TableIdent: table.Ident{
+			Name:   model.NewCIStr(mysql.TablePrivTable),
+			Schema: model.NewCIStr(mysql.SystemDB),
+		},
+		ColNames: []string{"Host", "User", "DB", "Table_name", "Table_priv", "Column_priv"},
+	}
+	values := make([][]expression.Expression, 0, 1)
+	value := make([]expression.Expression, 0, 3)
```
Comment (c4pt0r) [inline stmt/stmts/grant.go]: as mentioned above.

Comment (shenli) [inline stmt/stmts/grant.go]: PTAL

**Comment by shenli** [comment]
@c4pt0r PTAL

**Comment by ngaut** [comment]
@coocood PTAL

File: stmt/stmts/grant_test.go [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+	for _, v := range mysql.AllDBPrivs {
+		sql := fmt.Sprintf("SELECT %s FROM mysql.DB WHERE User=\"testDB1\" and host=\"localhost\" and db=\"test\";", mysql.Priv2UserCol[v])
+		tx = mustBegin(c, s.testDB)
+		rows, err := tx.Query(sql)
+		c.Assert(err, IsNil)
+		rows.Next()
+		var p string
+		rows.Scan(&p)
+		c.Assert(p, Equals, "Y")
+		c.Assert(rows.Next(), IsFalse)
```
Comment (siddontang) [inline stmt/stmts/grant_test.go]: maybe we can use matchRows to simplify codes.

Comment (shenli) [inline stmt/stmts/grant_test.go]: PTAL

File: stmt/stmts/grant.go [hunk 11]
```diff
@@ -0,0 +1,554 @@
...
+}
+
+// Find the schema by dbName.
+func (s *GrantStmt) getTargetSchema(ctx context.Context) (*model.DBInfo, error) {
+	dbName := s.Level.DBName
+	if len(dbName) == 0 {
+		// Grant *, use current schema
+		dbName = db.GetCurrentSchema(ctx)
+	}
+	if len(dbName) == 0 {
```
Comment (ngaut) [inline stmt/stmts/grant.go]: move up

File: stmt/stmts/grant.go [hunk 12]
```diff
@@ -0,0 +1,554 @@
...
+	sql := fmt.Sprintf(`SELECT Table_priv, Column_priv FROM %s.%s WHERE User="%s" AND Host="%s" AND DB="%s" AND Table_name="%s";`, mysql.SystemDB, mysql.TablePrivTable, name, host, db, tbl)
+	rs, err := ctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(ctx, sql)
+	if err != nil {
+		return "", "", errors.Trace(err)
+	}
+	row, err := rs.Next()
+	if err != nil {
+		return "", "", errors.Trace(err)
+	}
+	var (
```
Comment (c4pt0r) [inline stmt/stmts/grant.go]: var tPriv, cPriv string

File: stmt/stmts/grant.go [hunk 13]
```diff
@@ -0,0 +1,554 @@
...
+			v, ok := mysql.Priv2SetStr[p]
+			if !ok {
+				return "", errors.Errorf("Unknown column privilege %s", p)
+			}
+			if len(newColumnPriv) == 0 {
+				newColumnPriv = v
+			} else {
+				newColumnPriv = fmt.Sprintf("%s,%s", newColumnPriv, v)
+			}
+
```
Comment (ngaut) [inline stmt/stmts/grant.go]: Remove blank

File: stmt/stmts/grant.go [hunk 14]
```diff
@@ -0,0 +1,554 @@
...
+// Get current table scope privilege set from mysql.Tables_priv.
+// Return Table_priv and Column_priv.
+func getTablePriv(ctx context.Context, name string, host string, db string, tbl string) (string, string, error) {
+	sql := fmt.Sprintf(`SELECT Table_priv, Column_priv FROM %s.%s WHERE User="%s" AND Host="%s" AND DB="%s" AND Table_name="%s";`, mysql.SystemDB, mysql.TablePrivTable, name, host, db, tbl)
+	rs, err := ctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(ctx, sql)
+	if err != nil {
+		return "", "", errors.Trace(err)
+	}
+	row, err := rs.Next()
```
Comment (c4pt0r) [inline stmt/stmts/grant.go]: Need defer close?

File: stmt/stmts/grant.go [hunk 15]
```diff
@@ -0,0 +1,554 @@
...
+	col, ok := mysql.Priv2UserCol[priv]
+	if !ok {
+		return "", errors.Errorf("Unknown priv: %s", priv)
+	}
+	return fmt.Sprintf(`%s="Y"`, col), nil
+}
+
+// Compose update stmt assignment list for table scope privilege update.
+func composeTablePrivUpdate(ctx context.Context, priv mysql.PrivilegeType, name string, host string, db string, tbl string) (string, error) {
+	var (
```
Comment (c4pt0r) [inline stmt/stmts/grant.go]: var newTablePriv, newColumnPriv string

File: stmt/stmts/grant.go [hunk 16]
```diff
@@ -0,0 +1,554 @@
...
+}
+
+// Manipulate mysql.user table.
+func (s *GrantStmt) grantGlobalPriv(ctx context.Context, priv *coldef.PrivElem, user *coldef.UserSpecification) error {
+	asgns, err := composeGlobalPrivUpdate(priv.Priv)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	strs := strings.Split(user.User, "@")
+	userName := strs[0]
```
Comment (c4pt0r) [inline stmt/stmts/grant.go]: userName, host := strs[0], strs[1]

**Comment by c4pt0r** [comment]
I write some comments, rest LGTM.

**Comment by shenli** [comment]
PTAL @c4pt0r @ngaut

**Comment by c4pt0r** [comment]
LGTM

**Comment by ngaut** [comment]
LGTM

**Pull request merged**